integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-spark-service-bundle/nifi-spark-service/src/main/java/com/thinkbiganalytics/nifi/v2/spark/SparkJobserverService.java

```diff
@@ -0,0 +1,353 @@
+
+        boolean newContext = isNewContext(contextName);
+
+        if (newContext) {
+            Boolean contextRunning = checkIfContextExists(contextName);
+            if (!contextRunning) {
+                try {
+                    Map<String, String> params = new HashMap<>();
+                    params.put(ISparkJobServerClientConstants.PARAM_NUM_EXECUTORS, numExecutors);
+                    params.put(ISparkJobServerClientConstants.PARAM_SPARK_EXECUTOR_MEMORY, memPerNode);
+                    params.put(ISparkJobServerClientConstants.PARAM_NUM_CPU_CORES, numCPUCores);
+
+                    if (contextType != SparkContextType.SPARK_CONTEXT) {
+                        if (contextType == SparkContextType.SQL_CONTEXT) {
+                            params.put(ISparkJobServerClientConstants.PARAM_CONTEXT_TYPE, "spark.jobserver.context.SQLContextFactory");
+                        } else if (contextType == SparkContextType.HIVE_CONTEXT) {
+                            params.put(ISparkJobServerClientConstants.PARAM_CONTEXT_TYPE, "spark.jobserver.context.HiveContextFactory");
+                        }
+                    }
+
+                    if (async) {
+                        params.put(ISparkJobServerClientConstants.PARAM_SYNC, "false");
+                    } else {
+                        params.put(ISparkJobServerClientConstants.PARAM_SYNC, "true");
+                        params.put(ISparkJobServerClientConstants.PARAM_TIMEOUT, syncTimeout);
+                    }
+
+                    getLogger().info(
+                        String.format("Creating %s %s with %s executors, %s cores and %s memory per executor on Spark Jobserver %s", contextType, contextName, numExecutors, numCPUCores, memPerNode,
+                                      jobServerUrl));
+                    contextRunning = client.createContext(contextName, params);
+                    getLogger().info(String.format("Created %s %s on Spark Jobserver %s", contextType, contextName, jobServerUrl));
+
+                } catch (Exception ex) {
+                    getLogger().error(ex.getMessage());
+                } finally {
+                    synchronized (sparkContextsStates) {
+                        if (contextRunning) {
+                            sparkContextsStates.get(contextName).setStateAsRunning();
+                            sparkContextsStates.get(contextName).setTimeoutSeconds(contextTimeout);
+                        } else {
+                            sparkContextsStates.remove(contextName);
+                        }
+                        sparkContextsStates.notifyAll();
+                    }
+                }
+            }
+            return contextRunning;
+        } else {
+            return true;
+        }
+    }
+
+    @Override
+    public boolean deleteContext(String contextName) {
+
+        Boolean contextDeleted = !checkIfContextExists(contextName);
+
+        if (!contextDeleted) {
```

Comment:
What happens if another process deletes the context between the `checkIfContextExists` call and the `client.deleteContext` call? Should this method return `true` in that case or does in throw an exception?